### PR TITLE
Added `torch.baddbmm` operator

### LINKED ISSUE #326: Implement _VariableFunctionsClass.baddbmm of torch
## 🚀 Feature

Implement [`torch.baddbmm`](https://pytorch.org/docs/stable/generated/torch.baddbmm.html)

### Motivation

NeMo [Vision Transformer](https://github.com/NVIDIA/NeMo/blob/main/examples/vision/vision_transformer/megatron_vit_classification_pretrain.py)

cc @apaz-cli @tfogal

### PR DESCRIPTION
**Pull request by riccardofelluga**
<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

</details>

## What does this PR do?

This PR works towards supporting NeMo models, in particular fixes #326. I still think that the version mentioned in the issue is the inplace one, however this op is needed.

## Did you have fun?

One more op! 🚀

cc @apaz-cli @tfogal

### CODE REVIEW

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
...
+    utils.check_same_dtype(a, b1, b2)
+    utils.check(b1.ndim == 3, lambda: f"batch1 must be a 3D tensor, found {b1.ndim} instead.")
+    utils.check(b2.ndim == 3, lambda: f"batch2 must be a 3D tensor, found {b2.ndim} instead.")
+
+    if a.dtype not in dtypes.inexact_dtypes:
+        utils.check_type(beta, int)
+        utils.check_type(alpha, int)
+
+    t0 = matmul(b1, b2)
+    t1 = mul(alpha, t0)
```
Comment (nikitaved) [inline thunder/torch/__init__.py:3059]: no need to `mul` for the same reason.

File: thunder/torch/__init__.py [hunk 2]
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
 #
 
 
+@torchsymbol(torch.baddbmm, is_method=True)
+def baddbmm(
+    a: TensorLike, b1: TensorLike, b2: TensorLike, *, beta: float = 1.0, alpha: float = 1.0, out: TensorLike = None
+) -> TensorLike:
+    utils.check(out is None, lambda: "Non-None out is not supported", NotImplementedError)
+
+    utils.check_same_dtype(a, b1, b2)
```
Comment (nikitaved) [inline thunder/torch/__init__.py:3050]: We will probably do shortcuts below, so we should probably check that all things are on the same device.

File: thunder/tests/opinfos.py [hunk 3]
```diff
@@ -6153,6 +6153,63 @@ def tensor_1d_sample_generator(op, device, dtype, requires_grad, **kwargs):
 nn_ops = []
 
 
+def baddbmm_sample_generator(op, device, dtype, requires_grad, **kwargs):
+    make = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+    cases = (
+        ((0, 0, 0), (0, 0, 0), (0, 0, 0)),
+        ((3, 0, 5), (3, 0, 0), (3, 0, 5)),
+        ((0, 5, 6), (0, 5, 0), (0, 0, 6)),
+        ((3, 5, 6), (3, 5, 0), (3, 0, 6)),
```
Comment (nikitaved) [inline thunder/tests/opinfos.py:6162]: Let's add cases where the matmul part actually produces something other than zero

File: thunder/torch/__init__.py [hunk 4]
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
...
+    utils.check(b1.ndim == 3, lambda: f"batch1 must be a 3D tensor, found {b1.ndim} instead.")
+    utils.check(b2.ndim == 3, lambda: f"batch2 must be a 3D tensor, found {b2.ndim} instead.")
+
+    if a.dtype not in dtypes.inexact_dtypes:
+        utils.check_type(beta, int)
+        utils.check_type(alpha, int)
```
Comment (nikitaved) [inline thunder/torch/__init__.py:3056]: Could you please elaborate on these lines?

File: thunder/torch/__init__.py [hunk 5]
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
...
+
+    t0 = matmul(b1, b2)
```
[the remark this answers is not in the document]
Comment (riccardofelluga) [reply thunder/torch/__init__.py:3058]: That's true, however I've noticed that it broke the gradient computation. Let me see what I can do

File: thunder/torch/__init__.py [hunk 6]
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
 #
 
 
+@torchsymbol(torch.baddbmm, is_method=True)
+def baddbmm(
+    a: TensorLike, b1: TensorLike, b2: TensorLike, *, beta: float = 1.0, alpha: float = 1.0, out: TensorLike = None
```
Comment (mruberry) [inline thunder/torch/__init__.py:3046]: `, /, *,` let's not allow binding to names like `a`, `b1` and `b2`

File: thunder/torch/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mruberry) [reply thunder/torch/__init__.py:3058]: Curious what's going on with that; also fyi @IvanYashchuk

File: thunder/torch/__init__.py (hunk 1 above)
Comment (mruberry) [reply thunder/torch/__init__.py:3059]: nit: `alpha * t0`

File: thunder/torch/__init__.py [hunk 7]
```diff
@@ -3041,6 +3041,26 @@ def batch_norm(
...
+    utils.check(b2.ndim == 3, lambda: f"batch2 must be a 3D tensor, found {b2.ndim} instead.")
+
+    if a.dtype not in dtypes.inexact_dtypes:
+        utils.check_type(beta, int)
+        utils.check_type(alpha, int)
+
+    t0 = matmul(b1, b2)
+    t1 = mul(alpha, t0)
+
+    return add(t1, a, alpha=beta)
```
Comment (mruberry) [inline thunder/torch/__init__.py:3061]: nit: `t1 + (a * beta)`

**Comment by mruberry** [comment]
Looks really good so far, @riccardofelluga! Just some minor tweaks

File: thunder/torch/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tfogal) [reply thunder/torch/__init__.py:3058]: it is odd to break gradient...

but, i would think the matmul call itself would do this optimization if possible, and it'll do so with fewer instructions. If we want to do this, I'd recommend doing it at a higher level, via (say) a constant propagation pass, which could remove the op entirely and then potentially DCE some tensors.

[the remark this answers is not in the document]
Comment (mruberry) [reply thunder/torch/__init__.py:3058]: You're absolutely right that the logic could be handled by an optimization pass, @tfogal. I don't think we currently perform numeric simplifications. 

Also, is it true that we don't need to perform this matmul if `alpha` is zero? What if one or both of the tensors contained NaNs?

File: thunder/tests/opinfos.py [hunk 8]
```diff
@@ -6152,6 +6152,72 @@ def tensor_1d_sample_generator(op, device, dtype, requires_grad, **kwargs):
...
+        ((3, 5, 6), (3, 5, 8), (3, 8, 6)),
+    )
+
+    constants_cases = (
+        (2.0, 2.0),
+        (1.0, 0.0),
+        (0.0, 1.0),
+    )
```
Comment (IvanYashchuk) [inline thunder/tests/opinfos.py:6169]: `float('inf')`, `float('-inf')`, `float('nan')` are also important to check besides 0.0, 1.0, and 2.0. Could you please add them?

**Comment by riccardofelluga** [comment]
I've added the singularities in the testing. However, I reverted back to the code without the early exits because doing so would require writing the backward implementation for those computation paths. At the moment I think we can live with it and then in the future we might want to implement some optimization pass that does this for us.  

cc @nikitaved @mruberry @tfogal

**Review by mruberry — APPROVED** [review]

**Review by tfogal — APPROVED** [review]
